engine/access/rpc/backend/connection_factory.go

```diff
@@ -253,8 +294,106 @@ func getGRPCAddress(address string, grpcPort uint) (string, error) {
+// if the circuit breaker is disabled. If the response from the server indicates an unavailable status, it invalidates
+// the corresponding client.
+func (cf *ConnectionFactoryImpl) createClientInvalidationInterceptor(
+	address string,
+	clientType clientType,
+) grpc.UnaryClientInterceptor {
+	if cf.CircuitBreakerConfig == nil || !cf.CircuitBreakerConfig.Enabled {
+		clientInvalidationInterceptor := func(
+			ctx context.Context,
+			method string,
+			req interface{},
+			reply interface{},
+			cc *grpc.ClientConn,
+			invoker grpc.UnaryInvoker,
+			opts ...grpc.CallOption,
+		) error {
+			err := invoker(ctx, method, req, reply, cc, opts...)
+			if status.Code(err) == codes.Unavailable {
+				switch clientType {
+				case AccessClient:
+					cf.InvalidateAccessAPIClient(address)
+				case ExecutionClient:
+					cf.InvalidateExecutionAPIClient(address)
+				}
+			}
+
+			return err
+		}
+
+		return clientInvalidationInterceptor
+	}
+
+	return nil
+}
+
+// The simplified representation and description of circuit breaker pattern, that used to handle node connectivity:
+//
+// Circuit Open --> Circuit Half-Open --> Circuit Closed
+//      ^                                      |
+//      |                                      |
+//      +--------------------------------------+
+//
+// The "Circuit Open" state represents the circuit being open, indicating that the node is not available.
+// This state is entered when the number of consecutive failures exceeds the maximum allowed failures.
+//
+// The "Circuit Half-Open" state represents the circuit transitioning from the open state to the half-open
+// state after a configured restore timeout. In this state, the circuit allows a limited number of requests
+// to test if the node has recovered.
+//
+// The "Circuit Closed" state represents the circuit being closed, indicating that the node is available.
+// This state is initial or entered when the test requests in the half-open state succeed.
+
+// createCircuitBreakerInterceptor creates a client interceptor for circuit breaker functionality. It should only be
+// created if the circuit breaker is enabled. All invocations will go through the circuit breaker to be tracked for
+// success or failure of the call.
+func (cf *ConnectionFactoryImpl) createCircuitBreakerInterceptor() grpc.UnaryClientInterceptor {
+	if cf.CircuitBreakerConfig != nil && cf.CircuitBreakerConfig.Enabled {
+		circuitBreaker := gobreaker.NewCircuitBreaker(gobreaker.Settings{
+			// The restore timeout is defined to automatically return the circuit breaker to the HalfClose state.
+			Timeout: cf.CircuitBreakerConfig.RestoreTimeout,
+			ReadyToTrip: func(counts gobreaker.Counts) bool {
+				// The number of maximum failures is checked before the circuit breaker goes to the Open state.
+				return counts.ConsecutiveFailures >= cf.CircuitBreakerConfig.MaxFailures
+			},
+			MaxRequests: cf.CircuitBreakerConfig.MaxRetries,
```

Comment:
This naming is a little confusing. `MaxRequests` implies that it will allow up to (1 by default) requests. You've renamed it to `MaxRetries` which implies it should allow 2 requests, the initial and a single retry. Reading the doc, it seems like the library chose a confusing name. If that's the case, can you leave a comment here for future readers